insights/parsers/__init__.py

```diff
@@ -498,38 +506,90 @@ def keyword_search(rows, **kwargs):
-        matcher_fn = matchers['default']
-        if '__' in key:
-            key, matcher = key.split('__', 1)
+    txform_cache_attr = '_transform_cache'
+    if parent is None and hasattr(rows, '__dict__'):
+        parent = rows
+    # Uncomment this 'if' to check that all the parsers are supplying an
+    # object, somehow, that can store our transformed row cache.
+    # if parent is None:
+    #     print("Invoked with no parent arg on primitive container - use parent=self argument to cache row transform")
+    # The actual txkeys cache gets used rarely - in a basic test with a real
+    # archive it only got used eleven times.  But that's a saving...
+    if parent is not None and hasattr(parent, txform_cache_attr):
+        txkeys = getattr(parent, txform_cache_attr)
+    else:
+        # Store the translation from the search key to the key in the data.
+        # We start with the reverse transform - the key, to the transformed
+        # key.  This avoids doing the string replacement for every key in
+        # every row.  Then we invert that to map from the search term (with
+        # spaces and dashes replaced by underscores) to the key in the row
+        # data.
+        rev_txkeys = dict()
+        for row in rows:
+            for key in row.keys():
+                if key not in rev_txkeys:
+                    # Transform the term in a keyword_search() call into the
+                    # key in the rows
+                    rev_txkeys[key] = key.replace(' ', '_').replace('-', '_')
+        txkeys = dict(
+            (txkey, key)
+            for key, txkey in rev_txkeys.items()
+        )
+        if parent is not None:
+            setattr(parent, txform_cache_attr, txkeys)
+
+    # pre-compile the kwargs to find the matcher function and underlying key.
+    # Store these in a list of tuples for fast iteration and unpacking
+    search_terms = list()
+    for search_keyword, value in kwargs.items():
+        data_key = search_keyword
+        if '__' not in data_key:
+            matcher = 'equals'
+        else:
+            data_key, matcher = search_keyword.split('__', 1)
             if matcher not in matchers:
                 # put key back the way we found it, matcher fn unchanged
-                key = key + '__' + matcher
-            else:
-                matcher_fn = matchers[matcher]
-        return key in my_row and matcher_fn(my_row[key], value)
-
-    data = []
+                data_key = search_keyword
+                matcher = 'equals'
+        # If the data key sought is not in the row data, then we can say for
+        # sure that the search will never match...
+        if data_key not in txkeys:
+            return []
+        search_terms.append((
+            txkeys[data_key], matcher, matchers[matcher], value
+        ))
+
+    def key_match(data_key, matcher, matcher_fn, value):
+        if matcher == 'equals':
+            return data_key in row and row[data_key] == value
```

Comment:
CMIIW, I think the `matcher_fn` of `equals` (`lambda s, v: s == v`), would also work in the way of L588. Why not just remove this check and return L588 directly?
```python
def key_match(data_key, matcher, matcher_fn, value):
    return data_key in row and matcher_fn(row[data_key], value)
```

Reply:
I tested that and it was slower.  While it's nicer to have the single line of code, I made a judgement call that we do almost all of our comparisons using the `equals` matcher so making that as fast as possible was a priority.